docs/RFCS/20210519_bounded_staleness_reads.md

```diff
@@ -0,0 +1,912 @@
+This section discusses the implementation of bounded staleness reads from the
+perspective of the KV layer.
+
+### External Interface
+
+At a high-level, the KV client will expose a new `kv.BoundedStalenessSender`
+interface, exposed on a `kv.DB`, for upper levels to interact with bounded
+staleness reads. This interface will provide two methods.
+
+First, it will embed the `kv.Sender` interface, so it will expose a `Send`
+method which accepts a non-transactional, read-only `BatchRequest` with a
+`min_timestamp_bound` set in its `Header`. This `Sender` implementation will
+issue the `BatchRequest` while providing the [bounded staleness
+guarantee](#guarantees).
+
+Second, it will provide access to a `BoundedStalenessNegotiator` instance,
+which is capable of determining the `local_resolved_timestamp` for a set
+of spans.
+
+Together, the `BoundedStalenessSender` will looks like:
+```go
+// BoundedStalenessSender is a specialized Sender that is capable of
+// orchestrating bounded-staleness reads, given non-transactional, read-only
+// BatchRequests which set a min_timestamp_bound in their Header.
+type BoundedStalenessSender interface {
+    Sender
+    Negotiator() BoundedStalenessNegotiator
+}
+
+// BoundedStalenessNegotiator provides the local resolved timestamp for a
+// collection of key spans.
+type BoundedStalenessNegotiator interface {
+    LocalResolvedTimestamp(
+    	ctx context.Context, minTSBound hlc.Timestamp, spans []roachpb.Span,
+    ) (hlc.Timestamp, error)
+}
+```
+
+### Protobuf API Changes
+
+To support the `BoundedStalenessSender` implementation, we will make three
+changes to the KV Protobuf API.
+
+First, we will introduce a new `ReadConsistencyType` called `BOUNDED_STALENESS`.
+This `ReadConsistencyType` will return `false` from `RequiresReadLease`, meaning
+that requests with this consistency level set will be routed to the nearest
+replica for a range by the `DistSender`.
+
+Second, we will introduce a new field on the `BatchRequest` `Header` struct
+called `min_timestamp_bound`. This field can only be set when the header's
+`ReadConsistency` is set to `BOUNDED_STALENESS` and when the `timestamp` field
+is not set.
+
+Third, we will introduce a new `GetResolvedTimestamp` request and response pair.
+The intention is for this request type to be sent in `BatchRequests` with an
+`INCONSISTENT` consistency level so that it routes to the nearest replica, skips
+leaseholder checks, and does not wait in the replica's concurrency manager for
+latches. During evaluation, it computes the resolved timestamp over its key span
+on the replica that it is evaluating on. This means taking the minimum of the
+replica's closed timestamp and the timestamp before any locks in its key span.
+The `GetResolvedTimestamp` response type will implement the `combinable`
+interface and will take the minimum of any two responses.
+
+#### GetResolvedTimestamp efficiency
+
+`GetResolvedTimestamp` will initially be fairly expensive due to its scan for
```

Comment:
In the short term, we should also be able to leverage the MVCCStats. When `!ContainsEstimates && IntentCount == 0`, then there aren't any intents.